pkg/controller/garbagecollector/garbage_collector.go

```diff
@@ -0,0 +1,421 @@
+		attemptToOrphan:       attemptToOrphan,
+		registeredRateLimiter: NewRegisteredRateLimiter(deletableResources),
+		absentOwnerCache:      absentOwnerCache,
+	}
+	gb := &GraphBuilder{
+		metaOnlyClientPool:                  metaOnlyClientPool,
+		registeredRateLimiterForControllers: NewRegisteredRateLimiter(deletableResources),
+		restMapper:                          mapper,
+		graphChanges:                        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "garbage_collector_graph_changes"),
+		uidToNode: &concurrentUIDToNode{
+			RWMutex:   &sync.RWMutex{},
+			uidToNode: make(map[types.UID]*node),
+		},
+		attemptToDelete:  attemptToDelete,
+		attemptToOrphan:  attemptToOrphan,
+		absentOwnerCache: absentOwnerCache,
+	}
+	if err := gb.monitorsForResources(deletableResources); err != nil {
+		return nil, err
+	}
+	gc.dependencyGraphBuilder = gb
+
+	return gc, nil
+}
+
+func (gc *GarbageCollector) Run(workers int, stopCh <-chan struct{}) {
+	defer gc.attemptToDelete.ShutDown()
+	defer gc.attemptToOrphan.ShutDown()
+	defer gc.dependencyGraphBuilder.graphChanges.ShutDown()
+
+	glog.Infof("Garbage Collector: Initializing")
+	gc.dependencyGraphBuilder.startMonitors(stopCh)
+	if !cache.WaitForCacheSync(stopCh, gc.dependencyGraphBuilder.MonitorsSynced) {
+		return
+	}
+	glog.Infof("Garbage Collector: All resource monitors have synced. Proceeding to collect garbage")
+
+	// worker
+	go wait.Until(gc.dependencyGraphBuilder.runProcessGraphChanges, 1*time.Second, stopCh)
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(gc.runAttemptToDeleteWorker, 1*time.Second, stopCh)
+		go wait.Until(gc.runAttemptToOrphanWorker, 1*time.Second, stopCh)
+	}
+	Register()
+	<-stopCh
+	glog.Infof("Garbage Collector: Shutting down")
+}
+
+func (gc *GarbageCollector) runAttemptToDeleteWorker() {
+	for gc.attemptToDeleteWorker() {
+	}
+}
+
+func (gc *GarbageCollector) attemptToDeleteWorker() bool {
+	item, quit := gc.attemptToDelete.Get()
+	if quit {
+		return false
+	}
+	defer gc.attemptToDelete.Done(item)
+	n, ok := item.(*node)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expect *node, got %#v", item))
+		return true
+	}
+	err := gc.attempToDeleteItem(n)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Error syncing item %#v: %v", n, err))
+		// retry if garbage collection of an object failed.
+		gc.attemptToDelete.AddRateLimited(item)
+	}
+	return true
+}
+
+func objectReferenceToMetadataOnlyObject(ref objectReference) *metaonly.MetadataOnlyObject {
+	return &metaonly.MetadataOnlyObject{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: ref.APIVersion,
+			Kind:       ref.Kind,
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Namespace: ref.Namespace,
+			UID:       ref.UID,
+			Name:      ref.Name,
+		},
+	}
+}
+
+// classify the latestReferences to three categories:
+// solid: the owner exists, and is not "waiting"
+// dangling: the owner does not exist
+// waiting: the owner exists, its deletionTimestamp is non-nil, and it has
```

Comment:
Can we call this `waitingChildDeletion` or some such?  The code down in the loop where it was deciding to delete surprised me until I came back to the comment.